#286 Puzzle 213 in HttpFacade.java:66-69: Create new a class Facade

simonjenga opened the pull request:
#286 Puzzle 213 in HttpFacade.java:66-69: Create new a class Facade

static subclasses that encapsulate BaseFacade, please review

davvd commented:
@simonjenga Thanks, I will ask someone to do a code review here, thanks for your pull request

simonjenga commented:
@davvd @yegor256 Please assign someone to do a CR on this PR and provide some feedback.

davvd commented:
@krzyk please review, thanks

krzyk commented on .gitignore:
```diff
@@ -6,6 +6,7 @@ src/main/latex/*.pdf
 src/main/latex/*.aux
 src/main/latex/*.gz
 local-repo/
+/bin
```
@simonjenga why did you add this?

krzyk commented on s3auth-relay/src/main/java/com/s3auth/relay/BaseFacade.java:
```diff
@@ -0,0 +1,435 @@
+ * @since 0.1
+ */
+@ToString
+@EqualsAndHashCode(of = { "sockets", "server", "secured" })
```
@simonjenga `frontend` and `backend` is missing

krzyk commented on s3auth-relay/src/main/java/com/s3auth/relay/BaseFacade.java:
```diff
@@ -0,0 +1,435 @@
+    }
+
+    /**
+     * {@inheritDoc}
```
@simonjenga we don't need any javadoc on overriden methods

krzyk commented on s3auth-relay/src/main/java/com/s3auth/relay/BaseFacade.java:
```diff
@@ -0,0 +1,435 @@
+        final Runnable runnable = new VerboseRunnable(
+            new BaseFacade.ThreadRunnableDispatcher(thread), true, true
+        );
+        for (int idx = 0; idx < THREADS; ++idx) {
```
@simonjenga please add `BaseFacade.` before `THREADS` (all static fields have to be qualified with class name)

krzyk commented on s3auth-relay/src/main/java/com/s3auth/relay/FtpThread.java:
```diff
@@ -54,26 +54,24 @@
     /**
      * Queue of sockets to get from.
      */
-    @NotNull
```
@simonjenga why did you remove `NotNull`?

krzyk commented on s3auth-relay/src/main/java/com/s3auth/relay/FtpThread.java:
```diff
@@ -54,26 +54,24 @@
     /**
      * Hosts to work with.
      */
-    @NotNull
```
@simonjenga why did you remove `NotNull`?

krzyk commented on s3auth-relay/src/main/java/com/s3auth/relay/FtpThread.java:
```diff
@@ -84,10 +82,9 @@
      */
+    @Override
     public long dispatch() throws InterruptedException {
         assert this.hosts != null;
```
@simonjenga those asserts can be removed instead

simonjenga replied:
@krzyk These asserts will be romoved when the method will be implemented. They help to prevent warnings related to unused private fields in the class.

krzyk commented on s3auth-relay/src/main/java/com/s3auth/relay/FtpThread.java:
```diff
@@ -84,10 +82,9 @@
-     * Dispatch one request from the encapsulated queue.
-     * @return Amount of bytes sent to socket
-     * @throws InterruptedException If interrupted while waiting for the queue
+     * {@inheritDoc}
```
@simonjenga please remove this javadoc

krzyk commented on s3auth-relay/src/main/java/com/s3auth/relay/HttpThread.java:
```diff
@@ -136,13 +136,12 @@
-     * Dispatch one request from the encapsulated queue.
-     * @return Amount of bytes sent to socket
-     * @throws InterruptedException If interrupted while waiting for the queue
+     * {@inheritDoc}
```
@simonjenga pls remove this javadoc

krzyk commented on s3auth-relay/src/main/java/com/s3auth/relay/BaseFacade.java:
```diff
@@ -0,0 +1,435 @@
+        }
+
+        /**
+         * {@inheritDoc}
```
@simonjenga please remove this javadoc

krzyk commented on s3auth-relay/src/main/java/com/s3auth/relay/BaseFacade.java:
```diff
@@ -0,0 +1,435 @@
+        }
+
+        /**
+         * {@inheritDoc}
```
@simonjenga please remove this javadoc

krzyk commented on s3auth-relay/src/main/java/com/s3auth/relay/BaseFacade.java:
```diff
@@ -0,0 +1,435 @@
+    @ToString
+    @SuppressWarnings("PMD.DoNotUseThreads")
+    @Loggable(Loggable.DEBUG)
+    protected static final class FtpFacade extends BaseFacade {
```
@simonjenga plase make it a top level class

krzyk commented on s3auth-relay/src/main/java/com/s3auth/relay/BaseFacade.java:
```diff
@@ -0,0 +1,435 @@
+        }
+
+        /**
+         * {@inheritDoc}
```
@simonjenga please remove this javadoc

krzyk commented on s3auth-relay/src/main/java/com/s3auth/relay/BaseFacade.java:
```diff
@@ -0,0 +1,435 @@
+        }
+
+        /**
+         * {@inheritDoc}
```
@simonjenga please remove this javadoc

krzyk commented on s3auth-relay/src/main/java/com/s3auth/relay/BaseFacade.java:
```diff
@@ -0,0 +1,435 @@
+            final int sslport) throws IOException {
+            super(
+                2, "front", THREADS, "back",
+                        new ServerSocket(port),
```
@simonjenga this line (and two lines below) have to much indentation, it can be on the same level as `2`

krzyk commented on s3auth-relay/src/main/java/com/s3auth/relay/BaseFacade.java:
```diff
@@ -0,0 +1,435 @@
+    @ToString
+    @SuppressWarnings("PMD.DoNotUseThreads")
+    @Loggable(Loggable.DEBUG)
+    protected static final class HttpFacade extends BaseFacade {
```
@simonjenga please make it a top level class

krzyk commented on s3auth-relay/src/main/java/com/s3auth/relay/BaseFacade.java:
```diff
@@ -0,0 +1,435 @@
+        }
+
+        /**
+         * {@inheritDoc}
```
@simonjenga please remove this javadoc

krzyk commented on s3auth-relay/src/main/java/com/s3auth/relay/BaseFacade.java:
```diff
@@ -0,0 +1,435 @@
+       /**
+        * Base facade encapsulated.
+        */
+        private final transient BaseFacade facade;
```
@simonjenga why do you inherit AND encapsulate the base class?

krzyk commented on s3auth-relay/src/main/java/com/s3auth/relay/BaseFacade.java:
```diff
@@ -0,0 +1,435 @@
+     * @param socket The socket to report to
+     * @throws IOException If something wrong happens inside
+     */
+    protected void overflow(final Socket socket) throws IOException {
```
@simonjenga this method could be abstract if you would just inherit `BaseFacade`

krzyk commented on s3auth-relay/src/main/java/com/s3auth/relay/BaseFacade.java:
```diff
@@ -0,0 +1,435 @@
+     * in the subclasses as necessary.
+     */
+    protected void listen() {
+        // nothing to do, intentionally empty
```
@simonjenga this method couold be abstract if you would just inherit from `BaseFacade`

krzyk commented on s3auth-relay/src/main/java/com/s3auth/relay/RequestDispatcher.java:
```diff
@@ -0,0 +1,48 @@
+ * @version $Id$
+ * @since 0.1
+ */
+interface RequestDispatcher {
```
@simonjenga maybe this could be named just `Dispatchable`

krzyk commented on s3auth-relay/src/main/java/com/s3auth/relay/RequestDispatcher.java:
```diff
@@ -0,0 +1,48 @@
+package com.s3auth.relay;
+
+/**
+ * Dispatcher for a single {@link HttpThread} or {@link FtpThread}.
```
@simonjenga there is no need to mention `HttpThread` or `FtpThread`, interface is pretty generic

krzyk commented on s3auth-relay/src/main/java/com/s3auth/relay/BaseFacade.java:
```diff
@@ -0,0 +1,435 @@
+       /**
+        * Base facade encapsulated.
+        */
+        private final transient BaseFacade facade;
```
@simonjenga same here, why it is also encapsulated?

krzyk commented on s3auth-relay/src/main/java/com/s3auth/relay/BaseFacade.java:
```diff
@@ -0,0 +1,435 @@
+ */
+@ToString
+@EqualsAndHashCode(of = { "sockets", "server", "secured" })
+@SuppressWarnings({ "PMD.DoNotUseThreads", "PMD.TooManyMethods" })
```
@simonjenga when you move FTP and HTTP classes out, you can remove `TooManyMethods`

krzyk commented:
@simonjenga please see my comments

simonjenga commented:
@krzyk /bin is included in order to allow git to exclude compiled class files (*.class) from being commited. Eclipse compiles projects and puts those .class files in that bin folder.

I removed NotNull because validation has already been done at the constructor for those two fields. I don't think we need to duplicately validate them again. The constructor will never allow initialization of null values for those fields.

BaseFacade cannot be an abstract class. It will not allow encapsulation to be done. There is an error message related to Qulice when you try to do this.

#overflow(final Socket socket) and #listen() cannot be abstract methods since it would mean BaseFacade to be an abstract class which is not the case as explained above.

Base facade is encapsulated in order to access the private transient fields within it. Rememmber, we have been requested not to use getters/setters by the Architect. That is why i put those subclasses inside the Base class.

what is a top level class?

krzyk commented:
@simonjenga please respond in context (where the comments were)
As for the design, you could e.g. make `BaseFacade` (but you will have to find a better name) encapsulate object that will have specific behaviors (one for FTP and one for HTTP).

e.g. (suggestion, not actual design, you will have to try it out):
```
public class BaseFacade {
    private final Protocol protocol:

    public BaseFacade(Protocol prot) {
        this.protocol = prot;
    }

    private void overflow(Socket socket) {
        this.protocol.overflow(socket);
    }

....
}

class FTP implements Protocol {.....}
class HTTP implements Protocol {....}
```

simonjenga commented:
@krzyk While i try out to see if your proposed design will work, please look at this class ```com.jcabi.http.request.BaseRequest``` I have designed my solution with reference to it.

simonjenga commented:
@krzyk Not all ```{@inheritDoc}``` tags can be removed. Some are needed by qulice to prevent build errors.

Your proposed design to introduce an interface ```Protocol``` won't provide much help. We will still come across the same problems i experienced before in trying to access ```private transient``` fields and that is why our subclasses cannot be top level classes.

See the addressed code review changes.

krzyk commented:
@simonjenga as for `{@inheritDoc}` you can get rid of those, and change the checkstyle excludes to:
```
// @checkstyle SomeExclude (100 lines)
```
Just before method declaration, it will work there.

krzyk commented:
@simonjenga as for the fields, you can access them using methods, but they can't expose internal implementation.
Let's assume that you have a field to represent a `port`, you would name this field `prt`, and provide a method `port()` that will return this port as an int - this is OK because int is passed as a copy.
In case of objects you would first need to create a copy (for lists it would be something like `new ArrayList(oldList)`) or make sure the object is immutable (like `String`).

I don't know which fields you need to access, but try with the above.

simonjenga commented:
@krzyk I had originally implemented this solution with your suggested approach in the first attempt and it was rejected by the Architect. That is why in this second attempt, i put both subclasses as static inside BaseFacade in order to eliminate the need for ```getters```

https://github.com/yegor256/s3auth/pull/297#issuecomment-97934969
https://github.com/yegor256/s3auth/pull/297#issuecomment-100288152

The problem we have is that we do not know what the Architect's ideal solution is for the design is!

It's better if you could request for a merge in order to know if we are right from his review. It may also help to prevent lot's of rework.

krzyk commented:
@simonjenga let's do it this way, I will ask for reassignment of this ticket (as I don't like the design and proposing solutions would waste both our times) so someone else will have a look at it

krzyk commented:
@davvd please assign someone else

davvd commented:
> @davvd please assign someone else

@krzyk -15 points to your rating

davvd commented:
> @davvd please assign someone else

@krzyk no problem, I will try to find somebody else

simonjenga commented:
@krzyk I fully understand you. Some of these tickets have stringent coding requirements which you may not like or are not used to. Hope will meet in some other PR.

davvd commented:
@pinaf review this one, please

pinaf commented:
@davvd please assign someone else

pinaf commented:
@simonjenga sorry, I don't think I can help here

yegor256 commented:
@simonjenga you see, they are afraid of your big pull request :) I would recommend you to drop this one and create a smaller one, with less changes and more focused ones. It will be easier to review. Also, keep this article in mind: http://www.yegor256.com/2015/06/22/valid-reasons-to-reject-bug-fix.html and

davvd commented:
> @davvd please assign someone else

@pinaf 15 points was deducted from your rating

davvd commented:
> @davvd please assign someone else

@pinaf right, I will find someone else, no problem

simonjenga commented:
@yegor256 First it was about the design (No ```getters/setters```); and that was addressed even with a lack of communication. Now it's about a big! PR.

And mind you, they are not afraid of a about a big! PR. It's about the design requirements with this issue that ultimately leads to more time constraints while making tradeoffs. Issues that are a quick buck get reviewed rather fast :) 

Issue #286 is not small either and that is why this PR has so many changes to eliminate the duplication that is already there. I agree, changes in the ```javadocs``` should not have been done but now it's a little too late to undo them. Is that a reason enough to close this PR? Let me know...

I don't see how this PR can be made any smaller than it already is while fixing #286

davvd commented:
@caarlos0 please review, thanks

caarlos0 commented on s3auth-relay/src/main/java/com/s3auth/relay/BaseFacade.java:
```diff
@@ -0,0 +1,427 @@
+     * @param secure The secure server socket
+     * @checkstyle ParameterNumber (5 lines)
+     */
+    protected BaseFacade(final int frontthreads, final String frontprefix,
```
@simonjenga you could create an object with the number of threads and their name...

simonjenga replied:
@caarlos0 Once we create this object, it would mean we also have some ```getters``` in order to retrieve the number of threads and their name so as to pass them to this method ```Executors.newScheduledThreadPool(int corePoolSize, ThreadFactory threadFactory)```. We are restricted not to use ```getters```.

caarlos0 replied:
@simonjenga weird, but OK..

caarlos0 commented on s3auth-relay/src/main/java/com/s3auth/relay/BaseFacade.java:
```diff
@@ -0,0 +1,427 @@
+    private void process(final ServerSocket svr) {
+        final Socket socket;
+        try {
+            socket = svr.accept();
```
@simonjenga I think you forgot to close the socker

simonjenga replied:
addressed the issue

caarlos0 commented:
@simonjenga some notes, @krzyk did a great job reviewing so there is not much left for me...

simonjenga commented:
@caarlos0 I think it's better to leave the ```{@inheritDoc}``` tags as they are since suppressing the associated qulice warnings as instructed above by @krzyk leaves the method to look as follows which is more verbose as it currently is.
```
// @checkstyle RegexpMultilineCheck (100 lines)
// @checkstyle RegexpSinglelineCheck (100 lines)
/**
  * @param socket The socket to report to
  * @throws IOException If something wrong happens inside
  */ 
```

caarlos0 commented:
@simonjenga no, you can remove the entire javadoc, so the javadoc from the parent method will be used. You can do this for all `@Override` methods..

simonjenga commented:
@caarlos0 You're right, removing the entire ```javadoc``` is the best thing to do.
See the addressed code review changes.

caarlos0 commented:
@simonjenga nice, thanks!

caarlos0 commented:
@rultor merge

rultor commented:
> @rultor merge

@caarlos0 Thanks for your request. @yegor256 Please confirm this.

caarlos0 commented:
ping @yegor256

caarlos0 commented:
ping @yegor256

caarlos0 commented:
@davvd this task seems abandoned...

yegor256 commented:
@caarlos0 ok, closing

davvd commented:
> @davvd this task seems abandoned...

@caarlos0 @yegor256 will close it

caarlos0 commented:
@davvd closed already

davvd commented:
> @davvd closed already

@caarlos0 thanks, got it

davvd commented:
@caarlos0 thanks, paid, **23 mins** to your account, payment ID is `AP-29E52249P8937502L`, task took 1152 hours and 5 mins to complete

you're getting extra minutes here (c=8)

+23 to your rating, your total score is [+1388](http://www.netbout.com/b/35095?open=rating)
